spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala

```diff
@@ -511,7 +511,13 @@ trait OptimisticTransactionImpl extends TransactionalWrite
 
       val newProtocolForLatestMetadata =
         Protocol(readerVersionAsTableProp, writerVersionAsTableProp)
-      val proposedNewProtocol = protocolBeforeUpdate.merge(newProtocolForLatestMetadata)
+      // When the protocol versions are explicitly set on table features protocol we may
+      // downgrade to legacy protocol versions.
```

Comment:
I love a bit of color:
```suggestion
      // When the protocol versions are explicitly set on table features protocol we may
      // downgrade to legacy protocol versions. Legacy protocol versions can only be
      // used if a table supports *exactly* the set of features in that legacy protocol
      // version, with no "gaps". By merging in the protocol features from a particular
      // protocol version, we may end up with such a "gap-free" protocol. E.g. if a table
      // has only table feature "checkConstraints" (added by writer protocol version 3)
      // but not "invariants" and "appendOnly", then setting the minWriterVersion to
      // 2 or 3 will add "invariants" and "appendOnly", filling in the gaps for writer
      // protocol version 3, and then we can downgrade to version 3.  
```